src/systemdhook.c

```diff
@@ -642,42 +661,48 @@ static int prestart(const char *rootfs,
 		}
 	}
 
-	if (bind_mount(id, CGROUP_SYSTEMD, systemd_path, true)) {
+	if (bind_mount(id, CGROUP_SYSTEMD, systemd_path, false)) {
 		pr_perror("%s: Failed to bind mount %s on %s", id, CGROUP_SYSTEMD, systemd_path);
 		return -1;
 	}
 
 	/*
-	   Mount the writable systemd hierarchy into the container
+	   Mount the writable systemd hierarchies into the container
 	*/
-	_cleanup_free_ char *named_path = NULL;
-	named_path = get_process_cgroup_subsystem_path(id, pid, "name=systemd");
-	_cleanup_free_ char *systemd_named_path = NULL;
-	if (asprintf(&systemd_named_path, "%s/%s", rootfs, named_path) < 0) {
-		pr_perror("%s: Failed to create path for %s/%s", id, rootfs, systemd_named_path);
-		return -1;
-	}
-	if (bind_mount(id, named_path, systemd_named_path, false)) {
-		pr_perror("%s: Failed to bind mount %s on %s", id, CGROUP_SYSTEMD, systemd_named_path);
-		return -1;
-	}
 
-	/***
-	* chown will fail on /var/lib/docker files as they are not on the
-	* container so let's pass false to not have it done in the chperm
-	* function.
-	***/
-	if (chperm(id, systemd_named_path, mount_label, uid, gid, false) < 0) {
-		return -1;
-	}
+	int i;
+	for (i=0; i < sysdHierArraySize; i++) {
+		_cleanup_free_ char *named_path = NULL;
+		named_path = get_process_cgroup_subsystem_path(id, pid, sysdHierArray[i]);
+		_cleanup_free_ char *systemd_named_path = NULL;
+		if (asprintf(&systemd_named_path, "%s/%s", rootfs, named_path) < 0) {
```

Comment:
I think  you might have to free this each time through the loop, except the last one will be freed on exit, or move the definition of systemd_named_path inside the loop.